[IRGen] Noncopyable enums dont forward.

nate-chandler opened the pull request:
If a payload is non-copyable, do not use the forward-to-payload strategy for single-payload enums.

rdar://123197700

nate-chandler commented:
@swift-ci please test

jckarter commented on lib/IRGen/GenEnum.cpp:
```diff
@@ -2940,11 +2940,32 @@ namespace {
+          emitPayloadDestroy(*payload.ti);
+          IGF.Builder.CreateBr(exitBlock);
+
+          IGF.Builder.emitBlock(exitBlock);
```
This is duplicating the destruction codegen we should already emit if we pick the `Normal` CopyDestroyStrategy. I think it would be better to choose the correct CopyDestroyStrategy to begin with.

nate-chandler committed:
[IRGen] Noncopyable enums dont forward.

If a payload is non-copyable, do not use the forward-to-payload strategy
for single-payload enums.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test macosx platform

nate-chandler commented:
@swift-ci please clean test macosx platform

nate-chandler commented:
@swift-ci please clean test macos platform

nate-chandler commented:
@swift-ci please smoke test macos platform
